nimble/host/src/ble_gap.c

```diff
@@ -737,6 +737,30 @@ ble_gap_slave_reset_state(uint8_t instance)
 #endif
 }
 
+/**
+ * Clears advertisement and discovery state.  This function is necessary
+ * when the controller loses its active state (e.g. on stack reset).
+ */
+void
+ble_gap_master_slave_reset_state(void)
+{
+#if NIMBLE_BLE_ADVERTISE
+#if MYNEWT_VAL(BLE_EXT_ADV)
+    uint8_t i;
+
+    for (i = 0; i < BLE_ADV_INSTANCES; i++) {
+        ble_gap_slave_reset_state(i);
+    }
+#else
+    ble_gap_slave_reset_state(0);
+#endif
+#endif
+
+#if NIMBLE_BLE_SCAN
+    ble_gap_master_reset_state();
```

Comment:
For the master role, imho we need bit more:
1) when BLE_GAP_OP_M_DISC is ongoing, we need to call  ble_gap_disc_complete()
2) when BLE_GAP_OP_M_CONN is ongoing we probably should call ble_gap_master_connect_failure()

Reply:
1. Yes, we should call ` ble_gap_disc_complete`, have made changes.
2. Yes, however in `ble_hs_reset`, we also call `ble_gap_conn_broken` which ultimately calls GAP event `BLE_GAP_EVENT_DISCONNECT` to notify app. But I do understand, you want to cover the cases where  there may be `connection establishment procedure` and may not be any `active connection (established)`.

Reply:
ad 2. Because application started connection procedure and is waiting for a result of it. 
As the example, if device is not around and application sent Connect, host has internal timeout for Connection procedure. When it timeouts, then host do connect cancel and when connection complete arrives host is calling `ble_gap_master_connect_cancelled`. In our case we don't want to send connect cancel but we want to notify application that connection is failed.

Reply:
Thank you for the explanation :+1: